File: Data Structures & Algorithms/coin-change/submission-0.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int n = coins.length, dp[] = new int[amount + 1], INF = (int)(1e9);
        Arrays.fill(dp, INF);
        dp[0] = 0;

        for(int i = 0; i < n; ++i){
            for(int j = coins[i]; j <= amount; ++j){
                dp[j] = Math.min(dp[j], 1 + dp[j - coins[i]]);
            }
        }
        return dp[amount] == INF? -1 : dp[amount];
    }
}

File: Data Structures & Algorithms/coin-change/submission-0.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [10**9] * (amount + 1)
        dp[0] = 0

        for coin in coins:
            for val in range(coin, amount+1):
                dp[val] = min(dp[val], 1 + dp[val - coin])
        
        return dp[amount] if dp[amount] < (10**9) else -1
